six/modules/c++/six.sicd/tests/test_read_sicd_with_extra_des.cpp

```diff
@@ -0,0 +1,221 @@
+    secondHdr.setField("DESSHLPG", std::string("1", 125));
+    secondHdr.setField("DESSHLPT", std::string(" ", 20));
+    secondHdr.setField("DESSHLI", std::string(" ", 20));
+    secondHdr.setField("DESSHLIN", std::string(" ", 120));
+    secondHdr.setField("DESSHABS", std::string(" ", 200));
+    des.getSubheader().setSubheaderFields(secondHdr);
+
+    //Wrong length
+    nitf::TRE middleTRE(six::Constants::DES_USER_DEFINED_SUBHEADER_TAG,
+            "XML_DATA_CONTENT_283");
+    middleTRE.setField("DESCRC", "99999");
+    middleTRE.setField("DESSHFT", "XML00000");
+    middleTRE.setField("DESSHDT", "2016-06-01T00:00:00Z");
+    middleTRE.setField("DESSHRP", "1234567890123456789012345678901234567890");
+    middleTRE.setField("DESSHSI", std::string(" ", 60));
+    middleTRE.setField("DESSHSV", "Version 10");
+    middleTRE.setField("DESSHSD", "2016-06-01T00:00:00Z");
+    middleTRE.setField("DESSHTN", std::string(" ", 120));
+
+    nitf::DESegment middleDES = record.newDataExtensionSegment();
+    middleDES.getSubheader().getFilePartType().set("DE");
+    middleDES.getSubheader().getTypeID().set("XML_DATA_CONTENT");
+    middleDES.getSubheader().getVersion().set("01");
+    middleDES.getSubheader().getSecurityClass().set("U");
+
+    nitf::SegmentMemorySource middleSource(segmentData, strlen(segmentData),
+            0, 0, true);
+    mem::SharedPtr<nitf::SegmentWriter> middleSegmentWriter(new nitf::SegmentWriter);
+    middleSegmentWriter->attachSource(middleSource);
+    writer.addAdditionalDES(middleSegmentWriter);
+
+
+    //Wrong length again
+    nitf::TRE shortTRE(six::Constants::DES_USER_DEFINED_SUBHEADER_TAG,
+            "XML_DATA_CONTENT_005");
+    shortTRE.setField("DESCRC", "99999");
+
+    nitf::DESegment shortDES = record.newDataExtensionSegment();
+    shortDES.getSubheader().getFilePartType().set("DE");
+    shortDES.getSubheader().getTypeID().set("XML_DATA_CONTENT");
+    shortDES.getSubheader().getVersion().set("01");
+    shortDES.getSubheader().getSecurityClass().set("U");
+
+    nitf::SegmentMemorySource shortSource(segmentData, strlen(segmentData),
+            0, 0, true);
+    mem::SharedPtr<nitf::SegmentWriter> shortSegmentWriter(new nitf::SegmentWriter);
+    shortSegmentWriter->attachSource(shortSource);
+    writer.addAdditionalDES(shortSegmentWriter);
+
+    std::auto_ptr<TempFile> temp(new TempFile());
+    writer.save(&bandData[0], temp->pathname());
+    return temp;
+}
+}
+
+
+int main(int argc, char** argv)
+{
+    try
+    {
+        validateArguments(argc, argv);
+        const std::string xmlPathname(argv[1]);
+
+        six::XMLControlFactory::getInstance().addCreator(
+            six::DataType::COMPLEX,
+            new six::XMLControlCreatorT<six::sicd::ComplexXMLControl>());
+
+        std::auto_ptr<TempFile> nitf = createNITFFromXML(xmlPathname);
+        six::NITFReadControl reader;
+        reader.load(nitf->pathname());
+        std::cout << "Test passed\n";
```

Comment:
This is done.

I ran into a confusing comment while I was fixing this. In ReadControl.h, line 100-103, it claims “This container is not allocated, nor deleted by this class.” NITFReadControl both allocates and deletes the container. Is this meant to mean that subclasses are not required to allocate or delete the container? Or should I just change the comment?

From: Adam Sylvester [mailto:notifications@github.com]
Sent: Thursday, June 09, 2016 8:25 PM
To: ngageoint/six-library <six-library@noreply.github.com>
Cc: Jonathan Means <Jonathan.Means@mdaus.com>; Mention <mention@noreply.github.com>
Subject: Re: [ngageoint/six-library] Handling extra XML_DATA_CONTENT DESs when reading (#78)


In six/modules/c++/six.sicd/tests/test_read_sicd_with_extra_des.cpp<http://cp.mcafee.com/d/k-Kr6x0SyMU--yYCesdTdIIczxNEVvvdLzCm6kjhOqejrb38UsqenT3qabaarVEVsovdydorcDIel_ok1iG6JQVsSGxHtendImuLHUV_HYUyMYe79ZuVtdBV_AkkjhOYCeLR4kRHFGTpsVkffGhBrwqrhdICXYyUqenAkmkXL6XCOsVHkiP2cFASO7bV0zF7WOxfai3OVYy4tymNpIY_1PM68iFpKB9lwaNssvood7aoUT2Ivmfb0EzapdIxOVJNxNwS2_id41FrwhQzZpgDB91Vs-h2eNb6BQQg1vo87Owhd47xYUWq8066MGaxMw2h0l4wM30id0M4Cy0e6xyxEwc4_wq88_G4E4jh0qa0c7f6yASYOrLJRn>:

> +

> +int main(int argc, char** argv)

> +{

> +    try

> +    {

> +        validateArguments(argc, argv);

> +        const std::string xmlPathname(argv[1]);

> +

> +        six::XMLControlFactory::getInstance().addCreator(

> +            six::DataType::COMPLEX,

> +            new six::XMLControlCreatorT<six::sicd::ComplexXMLControl>());

> +

> +        std::auto_ptr<TempFile> nitf = createNITFFromXML(xmlPathname);

> +        six::NITFReadControl reader;

> +        reader.load(nitf->pathname());

> +        std::cout << "Test passed\n";

I think this is a good test but one thing is that you're just verifying reading didn't error out... not checking that we actually read the right thing out. So after you load(), if you can check that the ComplexData looks as expected, you should be good.

Note that you don't want to do just a string comparison of the XML because Linux and Windows will format things differently unless you have a different XML file for each platform. But you should be able to use some of those six::sicd::Utilities to quickly go from XML string to ComplexData and then use the equality operator to compare.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<http://cp.mcafee.com/d/2DRPoA820O76Qm77TQnANPxKVJBxAsed7bXVJYsOMOyqejhOrpop73zhO-Urhhphjvd7bz3VIhH3pAZxOLX2walgRKDbCRkdrFOVJyPRZv7fZvD4m7xMVfHTbFILfYyyyqenANR-EyCJtdmXbDaxVZicHs3jqpJATvAn3hOYyyODtUTsSjDdqymohBcCSgVv84t8_mk9VigunfAgzIiSbdDDUeuNSh_QN1BVBMQsKSjsju7e8Tjhtd78WUV2OtXnh7ef6zDQ4m6206hssvood7aoUT2Ivmfb0EzapdIxOVJNxNwS2_id41FrwhQzZpgDB91Vs-h2eNb6BQQg1vo87Owhd47xYUWq8066MGaxMw2h0l4wM30id0M4Cy0e6xyxEwc4_wq88_G4E4jh0qa0c7f6yASYOrL5uGqlCSs7oQ>, or mute the thread<http://cp.mcafee.com/d/2DRPowrhosvvhuj7e6XCSm6hMUQsLLCTNPb3a9EVd79JBxAsed7bXxJ55B5dYQsKcfCN6IdCjS7a_Ia0Fl3mWsKrlgRKDbCSbfnRYs_R-shou73A-LsKCOY_Oaa9EVuj7nWyaqRQRrIKsG7DR8OJMddICSjt-hsd7bOabatTztPpesRG9px6kOrp3BYxmsDpmSsGMFzsadAamz8x-20K2N03xow0FEV1oGlfhQbxw1xP1oyWaTr0Fo90g1aNmzUI1g53wUzE4b0idI1J0KIodMH7RzOMa8OCjr8sKrsosodwLQzh0qmU4t8_mk9VigunfAgzIiNFtd40nS21YE4jh1UveeCy01xIayEs80Ag5h8c0M4zgc19Ew3xEoEq831fU6y2fWxa14Qg6yw31PNEFdLcCNguPECb3sEM5>.


The information contained in this communication is confidential, is intended only for the use of the recipient(s) named above, and may be legally privileged. If the reader of this message is not the intended recipient, you are hereby notified that any dissemination, distribution, or copying of this communication is strictly prohibited. If you have received this communication in error, please re-send this communication to the sender and delete the original message or any copy of it from your computer system.